posthog/caching/insights_api.py

```diff
@@ -1,46 +1,112 @@
+) -> Tuple[bool, timedelta]:
+    """Return whether the insight should be refreshed now, and what's the minimum wait time between refreshes.
+
+    If a refresh already is being processed somewhere else, this function will wait for that to finish (or time out).
+    """
+    now = datetime.now(tz=pytz.timezone("UTC"))
     filter = get_filter(
         data=insight.dashboard_filters(dashboard_tile.dashboard if dashboard_tile is not None else None),
         team=insight.team,
     )
 
     target: Union[Insight, DashboardTile] = insight if dashboard_tile is None else dashboard_tile
     cache_key = calculate_cache_key(target)
-    caching_state = InsightCachingState.objects.filter(team_id=insight.team.pk, cache_key=cache_key, insight=insight)
-
-    refresh_frequency = DEFAULT_CLIENT_INSIGHT_ALLOWED_REFRESH_FREQUENCY
+    # Most recently queued caching state
+    caching_state = (
+        InsightCachingState.objects.filter(team_id=insight.team.pk, cache_key=cache_key, insight=insight)
+        .order_by("-last_refresh_queued_at")
+        .first()
+    )
 
     delta_days: Optional[int] = None
-
     if filter.date_from and filter.date_to:
         delta = filter.date_to - filter.date_from
         delta_days = ceil(delta.total_seconds() / timedelta(days=1).total_seconds())
 
-    if (hasattr(filter, "interval") and filter.interval == "hour") or (delta_days is not None and delta_days <= 7):
-        refresh_frequency = timedelta(minutes=3)
+    refresh_frequency = BASE_MINIMUM_INSIGHT_REFRESH_INTERVAL
+    if is_shared:
+        # The interval is longer for shared insights/dashboards
+        refresh_frequency = INCREASED_MINIMUM_INSIGHT_REFRESH_INTERVAL
+    elif getattr(filter, "interval", None) == "hour" or (delta_days is not None and delta_days <= 7):
+        # The interval is shorter for short-term insights
+        refresh_frequency = REDUCED_MINIMUM_INSIGHT_REFRESH_INTERVAL
+
+    refresh_insight_now = refresh_requested_by_client(request) and (
+        caching_state is None
+        or caching_state.last_refresh is None
+        or (caching_state.last_refresh + refresh_frequency <= now)
+    )
+
+    if refresh_insight_now:
+        has_refreshed_somewhere_else = _sleep_if_refresh_is_running_somewhere_else(caching_state, now)
+        if has_refreshed_somewhere_else:
+            refresh_insight_now = False
+
+    return refresh_insight_now, refresh_frequency
+
+
+def _sleep_if_refresh_is_running_somewhere_else(caching_state: Optional[InsightCachingState], now: datetime) -> bool:
+    """Prevent the same query from running concurrently needlessly."""
+    is_refresh_currently_running = _is_refresh_currently_running_somewhere_else(caching_state, now)
+    if is_refresh_currently_running:
+        assert caching_state is not None  # Isn't None due to condition in _is_refresh_currently_running_somewhere_else
+        while is_refresh_currently_running:
+            sleep(1)
```

Comment:
I guess this blocks Django so the traffic could bring down the web server by exhausting all the threads in waiting, but before this they'd have been hitting ClickHouse too so 🤷

Reply:
Clickhouse's number of parallel queries is smaller than what nginx/django can handle. Currently if CH is over capacity, all API requests that make queries will throw, but the rest (postgres-based stuff) will still work fine. After this, there's a danger that no requests will work if enough people load that one query at the same time 🤔